codex-rs/tui/src/onboarding/welcome.rs

```diff
@@ -3,23 +3,53 @@ use ratatui::layout::Rect;
 use ratatui::prelude::Widget;
 use ratatui::style::Stylize;
 use ratatui::text::Line;
+use ratatui::widgets::Paragraph;
 use ratatui::widgets::WidgetRef;
+use ratatui::widgets::Wrap;
 
+use crate::frames::FRAME_TICK_DEFAULT;
+use crate::frames::FRAMES_DEFAULT;
 use crate::onboarding::onboarding_screen::StepStateProvider;
+use crate::tui::FrameRequester;
 
 use super::onboarding_screen::StepState;
+use std::time::Duration;
+use std::time::Instant;
+
+const FRAME_TICK: Duration = FRAME_TICK_DEFAULT;
 
 pub(crate) struct WelcomeWidget {
     pub is_logged_in: bool,
+    pub request_frame: FrameRequester,
+    pub start: Instant,
 }
 
 impl WidgetRef for &WelcomeWidget {
     fn render_ref(&self, area: Rect, buf: &mut Buffer) {
-        let line = Line::from(vec![
-            ">_ ".into(),
-            "Welcome to Codex, OpenAI's command-line coding agent".bold(),
-        ]);
-        line.render(area, buf);
+        self.request_frame.schedule_frame_in(FRAME_TICK);
+
+        let frames = &FRAMES_DEFAULT;
+        let idx = if FRAME_TICK.as_millis() > 0 {
+            let steps =
```

Comment:
I don't believe this is guaranteed to render the frames in order?

Reply:
I think this is trying to maintain a consistent mapping of frames / time elapsed, but it is possible that frames get dropped. Maybe that's OK?